src/graph/expression_operators.cpp

```diff
@@ -55,8 +55,15 @@ Expr operator-(Expr a) {
   return Expression<NegNodeOp>(a);
 };
 
+Expr softmax(Expr a) {
+  return Expression<SoftmaxNodeOp>(a);
+}
+
+// @TODO: maybe get rid of this entirely to not obfuscate, what's going on inside.
+// @TODO: switch to log-masking everywhere?
 Expr softmax(Expr a, Expr mask) {
-  return Expression<SoftmaxNodeOp>(a, mask);
+  auto logMask = (1 - mask) * -99999999.f;
```

Comment:
It might be a bit more expensive, but I am quite sure it's negligible. Those masks are usually small as they broadcast to batch or time dimension and also computed once per batch. Even it enters this step once for every time step, the hashing algorithm will reuse exactly the same log-mask that was computed once. 

The pain the previous solution was inflicting on me is not worth it :)